graph/graph_structure.js

```diff
@@ -0,0 +1,174 @@
+        if(this.goal < 0){
+            console.log('No can do...set a goal first');
+            return(-1)
+        }
+        if(queue.length === 0) {
+            this.visited = [];
+            let t = depth ? 'Depth' : 'Breadth';
+            console.log('');
+            console.log('STARTING SEARCH: '+ t);
+            queue = [[start]];
+        }
+        steps += 1;
+        let to_extend = queue.shift();
+        let additions = this.extend(to_extend, depth, useH);
+        for(let p of additions) {
+            if (!depth || p.includes('TERM')) {
+                queue.push(p);
+            } else {
+                queue.unshift(p);
+            }
+            if (p.includes(this.goal)) {
+                console.log('Path found: ' + p);
+                console.log('Required ' + steps + ' steps.');
+                console.log('Queue: ');
+                console.log(queue)
+                return (queue);
+            }
+        }
+        return(this.search(start, depth, useH, queue, steps))
+    }
+    extend(path=[], depth, useH) {
+        // console.log('Visited: '+this.visited)
+        let node = path.pop();
+        path.push(node);
+        if (node === 'TERM'){
+            return([path]);
+        }
+        let output = [];
+        let tmp = [];
+
+        for( let n of this.graph.get(node).values() ) {
+            let visit = depth ? node : n;
+            if (this.visited.includes(visit) && !path.includes(n)) {
+                tmp = [...path];
+                tmp.push('TERM');
+                output.push(tmp);
+                // console.log('Path ('+path+') has already been extended elsewhere, dropping from queue.')
+            } else if(!path.includes(n)) {
+                tmp = [...path]
+                tmp.push(n);
+                output.push(tmp);
+            }
+            if (visit === n) {
+                this.visited.push(visit);
+            }
+        }
+        this.visited.push(node);
+
+        if(!depth && useH){
+            console.log(output)
+        }
+
+        if (useH) {
+            let g = this.goal;
+            let c = this.columns;
+            let sorter = this.sorter;
+            output.sort(function(x, y){return sorter(x, y, g, c)});
+            if (!depth) {
+                output = output.slice(0, 2);
+            }
+        }
+        return(output);
+    }
+    sorter(x, y, g, c){
+        let ax = [...x];
+        let ay = [...y];
+        function heuristicDistance(i, j=-1) {
+            if (j < 0){
+                j = g;
+            }
+            let ic = convertLoc2Coord(i);
+            let jc = convertLoc2Coord(j);
+            let hd = 0;
+            for(let x = 0;x < 2; x++){
+                hd += Math.abs(ic[x] - jc[x])**2;
+            }
+            return Math.sqrt(hd);
+        }
+        function convertLoc2Coord(i) {
+            let x = i % c;
+            let y = Math.floor(i / c);
+            return [x, y];
+        }
```

Comment:
i'll need some more background to understand this sorter and what these functions being defined within it are doing

Reply:
This is what I was messaging you about earlier, we can do a CR tonight.